### DISCUSSION TITLE
How to use 15.4.0-experimental-stream-defer.1 @defer directive?

### DISCUSSION THREAD

**Discussion by gabbigum** [graphql/graphql-js#3291] [Q&A] [answered]
Hi, all. I'm not sure what steps I need to take in order to make @defer work. I've setup this bare minimum express-graphql server:
```
var express = require('express');
var { graphqlHTTP } = require('express-graphql');
var { buildSchema } = require('graphql');

// Construct a schema, using GraphQL schema language
var schema = buildSchema(`
  type Query {
    hello: Greeting
  }

  type Greeting {
    text: String
    textList: [String]
  }
`);

// The root provides a resolver function for each API endpoint
var root = {
   hello: () => {
       return { text: 'Hello World', textList: ['Hi', 'Wow'] } 
    },
};

var app = express();
app.use('/graphql', graphqlHTTP({
  schema: schema,
  rootValue: root,
  graphiql: true,
}));
app.listen(4040);
console.log('Running a GraphQL API server at http://localhost:4040/graphql');
```

And I am executing such query: 
```
hello {
  text
  ... on Greeting @defer {
    textList
  }
}

```
And I get such response: 

```
{}
```

What I am missing and how can I make it work? I know it's experimental feature and still in development that's why I am not expecting it to work perfect. Thanks in advance!

**Comment by saihaj** [comment, marked as answer]
For defer and stream to work you need a multipart response setup. Checkout @n1ru4l bleeding edge platground https://github.com/n1ru4l/graphql-bleeding-edge-playground/blob/be4ca50be9f99a67029e53070289398c12f82679/server/main.ts#L82

**Reply by gabbigum** [reply]
Hello, @saihaj Although the example is helpful, we are using express-graphql instead of graphql-helix. I found [experimental version of stream-defer for express-graphql](https://github.com/graphql/express-graphql/tree/v0.12.0-experimental-stream-defer.1) and started playing with it. A problem that I encountered straight away was some validation error:

```
"errors": [
    {
      "message": "Validation error of type UnknownDirective: Unknown directive defer @ 'Entities/someEntities'",
    }
...
]
```

Is there any valid example on how to use express-graphql stream-defer, because I couldn't find any? If there isn't one I am willing to contribute when I get it working. Thanks for your time!

**Reply by n1ru4l** [reply]
@gabbigum You can go to some of the first commits of the bleeding edge playground - However, I eventually;lly migrated away from express-graphql because it was a pain to handle. https://github.com/n1ru4l/graphql-bleeding-edge-playground/tree/27f4f05f57af2540b6934bacd52a7f5d3f3757d1

**Reply by saihaj** [reply]
> Is there any valid example on how to use express-graphql stream-defer, because I couldn't find any?

There isn’t one that I am aware of

>  If there isn't one I am willing to contribute when I get it working. 

that would be awesome!

**Reply by gabbigum** [reply]
I made some progress on the matter. Correct me if I'm wrong, but I think that by simply changing both dependencies of graphql and express-graphql to use the experimental @defer feature, we get all the functionality we need in order to do it (we register @defer in our schema and we also have the handler for defer). This led me to think that the client fetching the data might be the problem which in my case was GraphiQL, because I was getting strange error: 

```{
  "message": "Unexpected number in JSON at position 3",
  "stack": "SyntaxError: Unexpected number in JSON at position 3"
}
```

After trying to execute the deferred query via Postman I was able to get the result back, which wasn't deferred by the way. I just got the full response straight away which looked like this: 

```
------
Content-Type: application/json; charset=utf-8
Content-Length: 61

{"data":{"deferTest":{"text":"Peter Parker"}},"hasNext":true}

-----
Content-Type: application/json; charset=utf-8
Content-Length: 86

{"data":{"defferedText":"Took a long time, he?"},"path":["deferTest"],"hasNext":false}

-----

```

Later I found that GraphiQL actually supports @defer and @stream by default from this PR: https://github.com/graphql/graphiql/pull/1770 and we can experiment with it [here](https://deploy-preview-1770--graphiql-test.netlify.app/?query=%7B%0A%20%20isTest%0A%20%20stream%3A%20streamable(delay%3A%20500)%20%40stream(initialCount%3A%202)%20%7B%0A%20%20%20%20text%0A%20%20%7D%0A%20%20anotherStream%3A%20streamable(delay%3A%20200)%20%40stream(initialCount%3A%200)%20%7B%0A%20%20%20%20text%0A%20%20%7D%0A%7D%0A).
But when I test it in the experimental GraphiQL above it doesn't actually show how it patches the result. I just get the full response at once. I've tried it on Safari and Brave Browser just to be sure it's not some browser rendering shenanigan, but I don't think it is.

Are there any clients that might support @defer, and what recommendations can you give me guys? I really appreciate your work and thanks again.

**Reply by saihaj** [reply]
Don't believe there are any clients supporting it yet.
- Relay has open PR https://github.com/facebook/relay/pull/3121 
- Apollo client has planned for `v4` https://github.com/apollographql/apollo-client/issues/8184

**Reply by gabbigum** [reply]
Hi, again. I managed to setup my @defer example app, for both client side and server side. Basically the server side stays the same I haven't changed anything. I've tried using meros and fetch-multipart-graphql without any success, so I just tried to use plain node.js and see if I can make it work. I'm posting the code below. 

If you wish I could create a PR in the experimental defer/stream branch for express-graphql with example folder containing readme and both client/server examples, so people can experiment with it and it would be great since it'll be my first OSS contribution. Because as trivial as it seems for a person like me who didn't know javascript was a lot of trouble to get all of this working. 😄

```
import http from 'http';

const data = `query defer_test {
  deferTest {
    text
    ... on GraphQLDeferTest @defer {
      defferedText
    }
  }
}`

const options = {
  hostname: 'localhost',
  port: 4040,
  path: '/graphql',
  method: 'POST',
  headers: {
    'Content-Type': 'application/graphql',
    'Content-Length': Buffer.byteLength(data)
  }
};

const req = http.request(options, (res) => {
  console.log(`STATUS: ${res.statusCode}`);
  console.log(`HEADERS: ${JSON.stringify(res.headers)}`);
  
  res.setEncoding('utf8');

  res.on('data', (chunk) => {
    console.log(`BODY: ${chunk}`);
  });

  res.on('end', () => {
    console.log('No more data in response.');
  });
});

req.on('error', (e) => {
  console.error(`problem with request: ${e.message}`);
});

// Write data to request body
req.write(data);
req.end();
```

**Reply by saihaj** [reply]
Awesome! yes please PR your demo it would be very helpful.

**Reply by gabbigum** [reply]
I've made the pull request - https://github.com/graphql/express-graphql/pull/784. Thanks @saihaj.

**Reply by gabbigum** [reply]
Hello, again @saihaj. Can I have my pull request merged? I signed the CLA :) Thanks!

**Reply by saihaj** [reply]
I currently don't have rights to merge there but pinged people who have. I will try again. Sorry about that 😞
